### PR TITLE
Feat tvl_change (#6)

### PR DESCRIPTION
**Pull request by eric9060**
scrapy crawl idx-tvl-change

### CODE REVIEW

**Comment by Byronlee** [comment]
**When I run your code, the code keeps failing. Have you ever encountered this situation?**

**url: https://api.llama.fi/lite/protocols2**

image

**Maybe because it is ascii encoding, resulting in an error when decode json**

image

**When I switch the url to https://api.llama.fi/protocols， all is ok.**

**Comment by eric9060** [comment]
thanks for your feedback
that`s really weird
the link (https://api.llama.fi/lite/protocols2) is exactly what the webpage used
图片
图片
and this is the result
图片
(set the chg_filter to 20% since there is no protocols 7d change > 80% at this moment)

**Comment by eric9060** [comment]
I also tested the link https://api.llama.fi/protocols
it also works fine, and there`s more result like bitfinex, binance

I will update the code soon

File: crawlers/indicators/spiders/tvl_change_monitoring_of_top_defi_projects/tvl_change_spider.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import scrapy
```
Comment (Byronlee) [inline crawlers/indicators/spiders/tvl_change_monitoring_of_top_defi_projects/tvl_change_spider.py:1]: According to the experience and actual broadcast conditions, I made some optimizations, take a look

```python
# encoding: utf-8
import scrapy

from crawlers.utils import SpiderBase, rds
from jinja2 import Template
from crawlers.utils.group_alarm import catch_except
from crawlers.utils.humanize import humanize_float_en
import requests
import json
from defillama import DefiLlama

class TVLChange(SpiderBase):
    name = 'idx-tvl-change'

    url = 'https://api.llama.fi/lite/protocols2'
    start_urls = ['https://api.llama.fi/lite/protocols2']
    headers = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.99 Safari/537.36" }

    def start_requests(self):
       self.parse(requests.get(self.url, headers = self.headers))

    @catch_except
    def parse(self, response, **kwargs):
        data = response.json()['protocols']
        print(data)
        if len(data) < 1:
            return
        for prot in data:
             self.process(prot)


    def process(self, prot):
        '''
        1、Project TVL > $100M
        2、TVL 1d change > 20%(up or down) or  TVL 7d changes > 50%(up or down) or TVL 7d changes > 100%(up or down)
        '''
        if prot['tvl'] <= 100000000:
            return None

        change_1d = (prot['tvl'] - prot['tvlPrevDay']) / prot['tvlPrevDay'] * 100
        change_7d = (prot['tvl'] - prot['tvlPrevWeek']) / prot['tvlPrevWeek'] * 100
        change_30d = (prot['tvl'] - prot['tvlPrevMonth']) / prot['tvlPrevMonth'] * 100
        params = {
            'project_name': prot['name'],
            'category': prot['category'],
            'symbol': prot['symbol'],
            'url': prot['url'],
            'tvl': humanize_float_en(prot['url']),
            'change_1d': change_1d,
            'change_7d': change_7d,
            'change_30d': change_30d,
        }

        if change_1d > 20:
            params['interval'] = '1'
            params['interval_change'] = change_1d
        if change_7d > 50:
            params['interval'] = '7'
            params['interval_change'] = change_7d
        if change_30d > 100:
            params['interval'] = '30'
            params['interval_change'] = change_30d

        print(params)
        print(Template(self.alert_en_template()).render(params))
        print(Template(self.alert_cn_template()).render(params))

    def alert_en_template(self):
        return """
According to KingData monitoring, in the past {{interval}} days, {{project_name}}'s TVL {{% if interval_change > 0%}} increased {{% else %}} decreased {{% endif %}} {{ interval_change }}%.

Project name: {{ project_name }}
Category: {{ category }}
Token: {{ symbol }}
Official Website: {{ url }}
Current TVL: {{ tvl }}
TVL(24H%): {{ change_1d }}%
TVL(7D%): {{ change_7d }}%
TVL(30D%): {{ change_30d }}%
"""

    def alert_cn_template(self):
        return """
据 KingData 监控，过去 {{interval}} 天, {{project_name}} 的 TVL {{% if interval_change > 0%}}上涨{{% else %}}下跌{{% endif %}} {{ interval_change }}%。

项目名称: {{ project_name }}
项目类别: {{ category }}
Token: {{ symbol }}
官网链接: {{ url }}
当前 TVL: {{ tvl }}
TVL(24H%): {{ change_1d }} 
TVL(7D%): {{ change_7d }}
TVL(30D%): {{ change_30d }}
"""
````

File: crawlers/indicators/spiders/tvl_change_monitoring_of_top_defi_projects/definition.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+6pm Daily
+
+## Indicator Alert logic (Required)
+
+1、Project TVL > $100M
+
+and
+
+2、TVL 7d changes > 80%(up or down)
```
Comment (Byronlee) [inline crawlers/indicators/spiders/tvl_change_monitoring_of_top_defi_projects/definition.md]: 2、TVL 1d change > 20%(up or down) or  TVL 7d changes > 50%(up or down) or TVL 7d changes > 100%(up or down)

File: crawlers/indicators/spiders/tvl_change_monitoring_of_top_defi_projects/definition.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+## Indicator Alert Demo (Required)
+
+EN
+
+According to KingData monitoring, the projects whose TVL fluctuated by more than 80% in the past 7 days:
```
Comment (Byronlee) [inline crawlers/indicators/spiders/tvl_change_monitoring_of_top_defi_projects/definition.md]: EN
```python
According to KingData monitoring, in the past {{interval}} days, {{project_name}}'s TVL {{% if interval_change > 0%}} increased {{% else %}} decreased {{% endif %}} {{ interval_change }}%.   

Project name: {{ project_name }}    
Category: {{ category }}   
Token: {{ symbol }}    
Official Website: {{ url }}   
Current TVL: {{ tvl }}   
TVL(24H%): {{ change_1d }}%   
TVL(7D%): {{ change_7d }}%   
TVL(30D%): {{ change_30d }}%   
```

CN

```python
据 KingData 监控，过去 {{interval}} 天, {{project_name}} 的 TVL {{% if interval_change > 0%}}上涨{{% else %}}下跌{{% endif %}} {{ interval_change }}%。   

项目名称: {{ project_name }}   
项目类别: {{ category }}   
Token: {{ symbol }}   
官网链接: {{ url }}  
当前 TVL: {{ tvl }}    
TVL(24H%): {{ change_1d }}   
TVL(7D%): {{ change_7d }}   
TVL(30D%): {{ change_30d }}  
```

**Review by Byronlee — COMMENTED** [review]
Optimized for experience

**Comment by eric9060** [comment]
@Byronlee sorry for the late update (just got COVID-19 these days and feel terrible)

图片

**Comment by Byronlee** [comment]
great job！

**Pull request merged**